### PR TITLE
A lot of dotty-specific changes to stdlib. [ci: last-only]

### PR DESCRIPTION
**Pull request by DarkDimius**
There are going to be a lot more. We need to have some procedure to backport those fixes to scala/scala.

@adriaanm Am I right in my assumption that it would be simpler for you if we have a single bulk PR for stdlib changes? We're not done yet and _I do not propose this PR for review_, but to discuss intention of merging changes to scala/scala and to showcase what changes are needed.

### CODE REVIEW

File: src/library/scala/collection/mutable/ListBuffer.scala [hunk 1]
```diff
@@ -69,7 +69,7 @@ final class ListBuffer[A]
 
   protected def underlying: List[A] = start
 
-  private def writeObject(out: ObjectOutputStream) {
+  private def writeObject(out: ObjectOutputStream) = {
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/ListBuffer.scala:72]: Need explicit annotation of Unit here (and in a bunch of other places).

File: src/library/scala/collection/TraversableOnce.scala [hunk 2]
```diff
@@ -402,8 +402,8 @@ trait TraversableOnce[+A] extends Any with GenTraversableOnce[A] {
 
 
 object TraversableOnce {
-  implicit def alternateImplicit[A](trav: TraversableOnce[A]) = new ForceImplicitAmbiguity
-  implicit def flattenTraversableOnce[A, CC[_]](travs: TraversableOnce[CC[A]])(implicit ev: CC[A] => TraversableOnce[A]) =
+  implicit def alternateImplicit[A](trav: TraversableOnce[A]): ForceImplicitAmbiguity = new ForceImplicitAmbiguity
```
Comment (Ichoran) [inline src/library/scala/collection/TraversableOnce.scala:405]: It's the return of Java boilerplate!  Are we really advocating this as best practice (explicit return type when you are literally just calling the constructor of that very type)?

Comment (adriaanm) [inline src/library/scala/collection/TraversableOnce.scala:405]: For implicits, yes, for some TBD value of "yes". Currently, type inference and implicit search cycles can cause very hard to understand implicit search ambiguities. If we can come up with some easy to understand scheme that softens this requirement, great! Otherwise, I'm on board with the change in principle.

**Comment by adriaanm** [comment]
Let's target 2.12.x for this, please. Also, could you include `[ci: last-only]` in the title of your 2.12 PR, while we're experimenting, to [avoid having the CI build every commit](https://github.com/scala/scabot/blob/master/README.md#pr-title-modifiers) until the PR is ready?

File: src/library/scala/Option.scala [hunk 3]
```diff
@@ -253,8 +253,8 @@ sealed abstract class Option[+A] extends Product with Serializable {
    *  @see map
    *  @see flatMap
    */
-  @inline final def foreach[U](f: A => U) {
-    if (!isEmpty) f(this.get)
+  final def foreach[U](f: A => U): Unit = {
```
Comment (adriaanm) [inline src/library/scala/Option.scala:256]: I don't think we should remove `@inline` here.

File: src/library/scala/math/Ordering.scala [hunk 4]
```diff
@@ -224,11 +224,7 @@ object Ordering extends LowPriorityOrderingImplicits {
   implicit object Unit extends UnitOrdering
 
   trait BooleanOrdering extends Ordering[Boolean] {
-    def compare(x: Boolean, y: Boolean) = (x, y) match {
-      case (false, true) => -1
-      case (true, false) => 1
-      case _ => 0
-    }
+    def compare(x: Boolean, y: Boolean) = if(x) { if (y) 0 else 1} else {if (y) -1 else 0}
```
Comment (adriaanm) [inline src/library/scala/math/Ordering.scala:227]: why is this change needed?

Comment (lrytz) [inline src/library/scala/math/Ordering.scala:227]: territory for the new optimizer! i'll be working on this hopefully next week.

Comment (sjrd) [inline src/library/scala/math/Ordering.scala:227]: +1! Scala.js can do it:

``` javascript
ScalaJS.s.s_math_Ordering$BooleanOrdering$class__compare__s_math_Ordering$BooleanOrdering__Z__Z__I = (function($$this, x, y) {
  if ((!x)) {
    if (y) {
      return (-1)
    }
  };
  if (x) {
    if ((!y)) {
      return 1
    }
  };
  return 0
});
```

Comment (lrytz) [inline src/library/scala/math/Ordering.scala:227]: 🏆 :-)

**Comment by soc** [comment]
As @adriaanm mentioned, please target 2.12.x. I'm happy to review!

File: src/library/scala/math/Ordering.scala (hunk 4 above)
Comment (DarkDimius) [inline src/library/scala/math/Ordering.scala:227]: >  why is this change needed?

It's not needed. It's just that I saw how inefficient the code was and I didn't resist the temptation to rewrite it. I'm fine with dropping this change if it's not welcome.

Comment (adriaanm) [inline src/library/scala/math/Ordering.scala:227]: In big changes like this, we like to separate concerns. The PR  should be restricted to changes that are crucial, and anything non-obvious must be motivated. Let's discuss whether to split up along module boundaries or along the kind of change. I think it makes most sense to have one commit per kind of change (procedure syntax, explicit result types for implicits, avoid name clashes for inner classes). I would also like to see a bytecode diff. We have automated this: https://github.com/retronym/libscala/tree/master/git-java (git diff invokes a custom command to generate only the essential diff). Here's an asm-based javap replacement that sorts by method name: https://github.com/adriaanm/scala/commit/afd15bb291b81dff4df4c22c384a4ab2c4714fad

**Comment by som-snytt** [comment]
You can use the code rewriting tool to remove procedure syntax.

Oh wait, DarkDimius _is_ the code rewriting tool.

**Comment by lrytz** [comment]
closing this, as it targets the wrong branch anyway

File: src/library/scala/math/Ordering.scala (hunk 4 above)
Comment (lrytz) [inline src/library/scala/math/Ordering.scala:227]: For the record, scalac can now also do it :) will add it as a test case. cfr-decompiler shows

```
    public int compare(boolean x, boolean y) {
        int n = false == x && true == y ? -1 : (true == x && false == y ? 1 : 0);
        return n;
    }
```

Comment (sjrd) [inline src/library/scala/math/Ordering.scala:227]: Nice!

Comment (adriaanm) [inline src/library/scala/math/Ordering.scala:227]: Indeed!

Comment (DarkDimius) [inline src/library/scala/math/Ordering.scala:227]: :+1:
